torchtune/utils/collate.py

```diff
@@ -133,3 +129,68 @@ def padded_collate_dpo(
     )
 
     return concatenated_input_ids, concatenated_labels
+
+
+def _padded_collate_packed(
+    sample: Dict[str, Any],
+    max_seq_len: int,
+    padding_idx: int = 0,
+    ignore_idx: int = CROSS_ENTROPY_IGNORE_IDX,
+) -> Dict[str, torch.Tensor]:
+    """Pad a batch of packed sequences to max sequence length in the batch, and
+    convert integer lists to tensors. Account for attention mask and position
+    ids.
+
+    This is a sample-wise collator and should not be used with the dataloader.
+
+    Sample should look like::
+        {
+            "tokens": List[int],
+            "labels": List[int],
+            "mask": Tensor,
+            "input_pos": List[int],
+        }
+
+    Args:
+        sample (Dict[str, Any]): A dictionary containing tokens, labels, mask,
+            and position ids
+        padding_idx (int): Padding index for input ids. Defaults to 0.
+        ignore_idx (int): Padding index for labels. Defaults to -100.
+
+    Returns:
+        Collated tokens, labels, mask, and position ids.
+    """
+    tokens = sample["tokens"]
+    labels = sample["labels"]
+    mask = sample["mask"]
+    input_pos = sample["input_pos"]
+
+    # Pad to max sequence length
+    tokens = F.pad(
+        torch.tensor(tokens), (0, max_seq_len - len(tokens)), value=padding_idx
+    )
+    labels = F.pad(
+        torch.tensor(labels), (0, max_seq_len - len(labels)), value=ignore_idx
+    )
+    # For attention mask, simply use identity matrix for the pad tokens
+    mask_pad = torch.eye(max_seq_len - mask.shape[0], dtype=torch.bool)
+    mask = torch.block_diag(mask, mask_pad)
+    # For position ids, continue to increment for pad tokens
+    input_pos_pad = torch.arange(
+        input_pos[-1] + 1, max_seq_len - len(input_pos) + input_pos[-1] + 1
+    )
+    # Do not go beyond max_seq_len - 1
+    input_pos_pad = input_pos_pad.clamp(max=max_seq_len - 1)
```

Comment:
I don't fully follow this. Does this only happen when `split_across_pack=True` or something? Also in that case where are we truncating the tokens and labels? Is it in the base dataset? If so it's a bit confusing to have it spread out across three places (though admittedly I don't really see a better way since base dataset obviously shouldn't care about input_pos)

Reply:
we are not truncating - the creation of the packs are "truncating" to a set max_seq_len. If we split across packs there is an edge case for the last sample if it bleeds over into the next pack where the pos ids continue to the end of that sample (say sample had length 5 and max_seq_len is 8, and [..., 0, 1, 2] ended up in the previous pack, and the sample got split into the last pack [3, 4]. We still have 6 slots left, how do we pad position ids here without going beyond max_seq_len?). So I just enforced this, these position ids should be ignored anyway by the loss. But the better solution is a padding mask